### ISSUE TITLE
depend-info in manifest mode

### ISSUE THREAD

**Issue by Taha-cmd** [microsoft/vcpkg#36520]
### Is your feature request related to a problem? Please describe.

I want to query the dependency graph for an entire project. The current `depend-info` command only works in classic mode, in which the dependencies are shown for only one port. To query the dependencies for an entire project, one would have to parse the `vcpkg.json` manifest and call `depend-info` for each port.

### Proposed solution

Provide an (explicit or implicit) manifest mode for the `depend-info` command.
Examples: 
* `vcpkg depend-info` <- search for vcpkg.json in current directory
* `vcpkg depend-info vcpkg.json`

**Comment by jimwang118** [comment]
You can create a new folder `newdir` in the `vcpkg/port` directory, and then copy your project's `vcpkg.json` to the new folder. Then run the following command to get all dependencies.
`./vcpkg depend-info newdir --format=tree`

**Comment by Taha-cmd** [comment]
That worked indeed. However, this is not really appropriate for automated build processes and pipelines scenario, as it would require modifying the vcpkg installation directory.

**Comment by jimwang118** [comment]
> That worked indeed. However, this is not really appropriate for automated build processes and pipelines scenario, as it would require modifying the vcpkg installation directory.

Yes, this is just a temporary solution, I will try to modify to read the vcpkg.json file directly to show the dependencies.

**Comment by dg0yt** [comment]
As a mitigation, you can try `vcpkg install --dry-run`.

**Comment by horenmar** [comment]
I'll note two things here
 * The simplest way of hacking around the current limitation is an overlay port: `vcpkg depend-info my-app --overlay-ports=.`
 * The information returned from the trick above can be wrong, because the versions it selects are not done according to the manifest, nor is it a simple checkout of the registries (we have a private vcpkg registry).

**Comment by j-hap** [comment]
Currently the only way to install a specific package version is via manifest mode. There is no way to define a package version on the classic command line install. There is none for depend-info either. To to see the dependencies of a specific package version, would be nice if depend-info either works with the manifest file or if there was a way to specify the package version on the command line.

**Comment by jimwang118** [comment]
@BillyONeal Can vcpkg query dependent packages in manifest mode?

**Comment by dg0yt** [comment]
> To to see the dependencies of a specific package version, would be nice if depend-info either works with the manifest file

What works is https://github.com/microsoft/vcpkg/issues/36520#issuecomment-1930449009.

**Comment by jimwang118** [comment]
Use the command `vcpkg install --dry-run` to achieve the results you need. I will close this issue. If you have new issue, please submit a new issue.

**Comment by dg0yt** [comment]
Please reopen.

**Comment by Samt43** [comment]
vcpkg install --dry-run does NOT permit to have parent/children relationship. This feature is needed.
All good package manager permit to do this.

**Comment by trifonov-borislav** [comment]
> vcpkg install --dry-run does NOT permit to have parent/children relationship. This feature is needed. All good package manager permit to do this.

Adding my vote for this

**Comment by reitowo** [comment]
Want this.

**Comment by mheyman** [comment]
vcpkg fixed the requirement for parsing the manifest, you can now just run `/path/to/vcpkg  depend-info --overlay-ports YOUR-VCPKG-OVERLAY-PORTS --vcpkg-root YOUR-VCPKG-ROOT --format tree` in the directory with your manifest.

The below is for historical value (or you are stuck on an older vcpkg for some reason).

----

With the help of `jq`, you can run `vcpkg depend-info` on a local `vcpkg.json`. The `jq` command parses the dependencies from the vcpkg.json file and formats them in a way that makes `vcpkg depend-info` happy. Subtleties include feature parsing and only putting in square brackets when there actually are features.

The following is in the order of ugliest to least ugly 😉 ...


Replace the `/path/to/vcpkg` and `YOUR-VCPKG-...` values with what makes sense in your environment.

### Windows Cmd
In do a `winget install jqlang.jq` to get `jq` then you can run a cmd shell line that looks like:
```batch
for /f "tokens=*" %a in ('jq -r "if .dependencies then [ .dependencies[] | if type==""string"" then . else ( .name + ( if ((.features // []) | length) > 0 then (""["" + (((.features // []) | join("",""))) + ""]"") else """" end ) ) end ] | join("" "") else empty end" vcpkg.json') do \path\to\vcpkg depend-info --overlay-ports YOUR-VCPKG-OVERLAY-PORTS --vcpkg-root YOUR-VCPKG-ROOT --format tree %a
```
You need to make it `%%a` instead of `%a` if you put this in a batch file.

### Windows Powershell
Using `jq` (get it like above):
```powershell
\path\to\vcpkg depend-info --overlay-ports YOUR-VCPKG-OVERLAY-PORTS --vcpkg-root YOUR-VCPKG-ROOT --format tree $(-split $(jq -r "if .dependencies then [ .dependencies[] | if type==\`"string\`" then . else ( .name + ( if ((.features // []) | length) > 0 then (\`"[\`" + (((.features // []) | join(\`",\`"))) + \`"]\`") else \`"\`" end ) ) end ] | join(\`" \`") else empty end" vcpkg.json))
```
Powershell's subexpressions produce one parameter so the `-split` is required to get multiple parameters passed into `vcpkg depend-info`.

_or_ as @sharadhr pointed out, no `jq` required:
```powershell
\path\to\vcpkg depend-info --overlay-ports YOUR-VCPKG-OVERLAY-PORTS --vcpkg-root YOUR-VCPKG-ROOT --format tree $(Get-Content -Raw vcpkg.json | ConvertFrom-Json | ForEach-Object { @($_.dependencies) | ForEach-Object { switch ($_) { { $_ -is [string] } { $_ } { $_.features -and $_.features.Count -gt 0 } { "$($_.name)[$($_.features -join ',')]" } default { $_.name } } } })
```

### Linux Bash
You get `jq` from your package manager (like `sudo apt install jq` on Ubuntu, etc...)
```bash
/path/to/vcpkg depend-info --overlay-ports YOUR-VCPKG-OVERLAY-PORTS --vcpkg-root YOUR-VCPKG-ROOT --format tree  $(jq -r 'if .dependencies then [ .dependencies[] | if type=="string" then . else ( .name + ( if ((.features // []) | length) > 0 then ("[" + (((.features // []) | join(","))) + "]") else "" end ) ) end ] | join(" ") else empty end' vcpkg.json)
```

**Comment by sharadhr** [comment]
On Windows, no need for a batch file or `jq`. In PowerShell:

```pwsh
Get-Content -Raw vcpkg.json | ConvertFrom-Json | ForEach-Object {
    @($_.dependencies) | ForEach-Object {
        switch ($_) {
            { $_ -is [string] } { $_ }
            { $_.features -and $_.features.Count -gt 0 } { "$($_.name)[$($_.features -join ',')]" }
            default { $_.name }
        }
    }
}
```
